RustLayout rework, Global Layout object

matejcik opened the pull request:
## Ladies and gentlemen, this is what you've been waiting for 

fixes #2299 
fixes #3280 

### Main features:

1. unified `ui.Layout` and `ui.ProgressLayout` classes
   * instead of per-model `RustLayout` that is kind-of-but-not-really a copypaste job
2. Global `ui.CURRENT_LAYOUT` that always unambiguously refers to the layout that is on the screen, or `None` if the screen is dark
3. Total overhaul of debuglink internals, leading, hopefully, to greater test stability and less races
   * `client.debug.wait_layout()` is now almost never needed, except in very special cases. `read_layout()` Does The Right Thing.
   * deadlock detection when you miss a ButtonAck
   * it "worked on my PC" but i'm sure you guys will figure out a way to break it anyway :woman_shrugging: 

### Minor niceties

* timeout for `loop.wait()` so you usually don't need to use `loop.race(xyz, loop.wait(...))` e8990ce5db4a138df0670e8fe79aec73b164f4bb
  * paving way to fixing #3117
* `loop.chan` is unused now. I left it in just in case but we can probably remove it 9f01c9f896e9268142e63a134642432579890a77
* nice simplification of `handle_session` now that it does not need to support debuglink 06cad6fa6635593bb4503dbda2d520bd71ecf523
* bug with `make test_emu_ui` in zsh not being random is fixed eb65cf669234f3ed1110a865d831dd08cd8d9b99
* `WipeDevice` will show a progress screen for the 2 seconds it takes to wipe storage. c98e54ba5ea1731e27dd56cba04e6ee4a03530f4
  * it's "stuck" because we can't animate, but at least the user is not staring at a black screen
* if a timeout kills your test, you'll see where in the input flow it froze 1a9d022da11f46953164faf789b5fad52e0d037e
* detection of invalid input flows that don't run all the way through fd914bd18f9cf3819b46229886780fd849d48f9d
* reduces chattiness of error output when your UI tests are failing (e15c4e84f0a250265440940f5a230a0e98d87304, same as #3665)

### Supersedes

#2335 
#3083 
#3665 

<!--
If you are a core dev:

Don't forget to set up the fields:
- assign yourself
- set priority to same as original issue
- add PR to sprint
- if Draft -> "in progress"
- if final PR -> "needs review"

If you're an external contributor, you can ignore this text.
-->

TychoVrahe committed:
refactor(core): expose USE_TOUCH and USE_BUTTON to uPy

[no changelog]

matejcik committed:
feat(core): introduce timeout to loop.wait()

matejcik committed:
refactor(core): remove unused features from loop.race

matejcik committed:
refactor(core): remove debuglink support from main handle_session

It is being moved into a custom handler for the debug app.

matejcik committed:
feat(core): add mailbox as a simplified version of chan

matejcik committed:
build(core): fix random order seed when using zsh

matejcik committed:
refactor(core/ui): improve LayoutObj initialization

* RequestPaint message is sent at construction time to force calculation
  of number of pages
* given that Attach corresponds to "start the layout" message, Child now
  responds to Attach the same way it responds to RequestPaint, by
  force-repainting everything.

matejcik committed:
fix(core/ui): repaint the title of Progress layout when requested

matejcik committed:
feat(core): show progress before starting to wipe device

This doesn't actually advance the progress while device erase is ongoing, but at least the user sees a loader.

matejcik committed:
docs(core): Layout lifecycle documentation

matejcik committed:
refactor(core): improve recovery_enter_share

matejcik committed:
tests: use read_layout everywhere

matejcik committed:
tests: fix small irregularities

matejcik committed:
fix(core): correctly repaint ShowMore component when requested

matejcik committed:
feat(core): drop support for after-step-hook

This was necessary for hooking display.refresh() with the old UI toolkit.
With the new one, we explicitly refresh the display after every paint, so
implicit after-step refresh seems no longer necessary.

matejcik committed:
style: fix complaining lints & formatters

matejcik commented:
### UI changes summary:

* all progress steps of all loaders are removed, we only screenshot the first time a loader is shown
  * except for "hold to lock", because the loader starts in progress already
* some "loading transaction" screens are gone
  * previously, a "loading transaction" screen would blink inbetween every Bitcoin confirmation dialog
  * now it only shows up if there is in fact a loading step (so e.g., between the last output and the total confirmation screen there is nothing happening to warrant a loader)
* some screens are now recorded that are immediately canceled
  * in `test_cancel`, in cancellation versions of various reset/backup/recovery flows
* `WipeDevice` now starts progress earlier

### Learn more

https://github.com/trezor/trezor-firmware/blob/matejcik/global-layout-only3/docs/core/misc/layout-lifecycle.md

### Reviewer instructions

I strongly recommend going by commits. There's a big one that does most of the things (b75dbe18e4c04a7c73f4c7ae9c5160d3cd17f4b8), the other ones are more reasonable.

Please also point out if you feel that something is not documented that should be.

matejcik commented:
TODO: this definitely wants a changelog entry

matejcik commented:
seems that there's a progress related piece of code that I didn't manage to commit. stay tuned for fixes.

matejcik commented:
also this does not feel like I added 11k of code, not sure what is going on...

matejcik committed:
docs(core): add changelog

obrusvit commented on docs/core/misc/layout-lifecycle.md:
Nice docu :+1:  A few immediate points:

- There is no information that `stop()` can lead to `STOPPED` state but in the function `def stop(...)` the documentation says that the next state can be `READY` or `STOPPED`
- I find it a little bit misleading that `stop()` might put state into `READY` and not `STOPPED`. Perhaps we can name the state `DONE` instead of `STOPPED`? 
- Place this document into `docs/SUMMARY.md` please.

matejcik committed:
refactor(core/ui): reusable timers

matejcik committed:
feat(tests): sync received responses

to prevent "unexpected response to Initialize" errors

matejcik committed:
fix(tests): restore test functionality on legacy

The global layout related changes were wrong for T1 where debuglink
behavior is significantly different; in particular, it is not always
possible to communicate over debuglink.

This change reverts to the old behavior for T1B1 and keeps the new one
only for core-based models.

matejcik committed:
feat(core): unify RustLayout, implement single global layout

matejcik committed:
fix(core): set default value for do_not_restart

otherwise if _handle_single_message raises an exception (which is
fortunately not something that should generally happen), the
finally-block would fail on "referenced before assignment"

matejcik committed:
fix(core): allow Ping while locked

same behavior as on Trezor One

matejcik committed:
fix(python): improve robustness of TrezorClientDebugLink setup

* improve sync_responses to work on uninitialized instance
* sync responses at construction time

matejcik committed:
chore(python): improve recovery type deprecation warning

matejcik committed:
fix(tests): stop waiting for background task after a timeout

Otherwise a test can hang if the result is not returned. This is not
even covered by pytest-timeout because if a test fails, the call to
task.kill() happens during teardown, and pytest-timeout doesn't cover
that.

matejcik committed:
style(core/rust): avoid unused warning

matejcik committed:
refactor(core/rust): introduce layout lifecycle states on Rust side

matejcik committed:
style: fix style problems

matejcik committed:
refactor(core/rust): create a single code path allowed to modify SwipeFlow state

matejcik committed:
fix(python): opportunistically catch a debuglink desync

In certain cases a DebugLinkState response can get stalled and debuglink
will not read it -- i.e., a situation analogous to wirelink de-sync.
There doesn't seem to be a good way to force-sync debuglink the same way
wirelink does it, but we can detect a wrong response to a
DebugLinkGetState and skip it.

matejcik committed:
chore(core/rust): add uDebug capability to more places

matejcik committed:
chore: update fixtures

matejcik committed:
fix(tests): do not trip deadlock detection in autolock test

matejcik committed:
fix(core): fix cancel behavior in SignMessage

matejcik committed:
fix(tests): do not place recorded.zip into the destdir

instead, download it to a temp location, as the Prophecy (and authors of
urlretrieve) foretold

matejcik committed:
fix(tests): correctly check that backup flow was aborted

matejcik committed:
fix(core): restore ellipsis functionality on TR

matejcik committed:
style(core/rust): fix linter complaints

matejcik committed:
chore: update fixtures

matejcik committed:
style: make style_check passing

mmilata commented on docs/core/misc/layout-lifecycle.md:
```diff
@@ -0,0 +1,196 @@
+# UI Layout Lifecycle
```
IIRC you mentioned that this is now outdated, though it doesn't seems so on the first look.

mmilata reviewed (commented):
Mostly nits, let's fix the backlight & tests and merge soon so it gets some testing before next release.

obrusvit reviewed (commented):
Some points, mostly nits.
In addition: `docs/event-loop.msd` could be updated with `mailbox` description.

matejcik commented on core/src/trezor/loop.py:
```diff
@@ -233,11 +240,41 @@ class wait(Syscall):
+        _step(self.task, __value)
+        return self._DO_NOT_RESCHEDULE
+
+    throw = send
```
this class implements the `Iterator` interface defined by methods `send()` `throw()` and `close()`, so I need all three. In fact in `_step` you can get `throw()` called.
but at the same time we handle `send()` and `throw()` the same way

obrusvit commented on core/embed/rust/src/ui/flow/swipe.rs:
```diff
@@ -296,23 +322,22 @@ impl SwipeFlow {
-#[cfg(feature = "micropython")]
-impl ObjComponent for SwipeFlow {
-    fn obj_place(&mut self, bounds: Rect) -> Rect {
+impl Layout<Result<Obj, Error>> for SwipeFlow {
```
Commentary above this `impl` is now outdated.

mmilata replied:
333c5ec

matejcik committed:
fix(tests): fix persistence tests, enable T3B1

matejcik commented on core/src/trezor/ui/layouts/common.py:
```diff
@@ -1,42 +1,98 @@
+    exc: ExceptionType = ActionCancelled,
+) -> Awaitable[None]:
+    action = interact(layout_obj, br_name, br_code, exc)
+    return action  # type: ignore ["UiResult" is incompatible with "None"]
```
we're not returning `UiResult` though, we are returning an `Awaitable`.
the semantics of function "raise if not confirmed" are "don't care about the result value" so we need to mask the fact that the returned Awaitable is technically returning a value.
(the other way would be to propagate the `UiResult` through sometimes rather deep call stacks -- which is ultimately pointless because those stacks don't actually care about the result value)

matejcik commented on core/src/apps/debug/__init__.py:
```diff
@@ -301,3 +350,89 @@ async def dispatch_DebugLinkOptigaSetSecMax(
+                # try to stay robust in such case.
+                log.exception(__name__, exc)
+
+    WORKFLOW_HANDLERS: dict[int, Handler] = {
```
it seemed this dict is small enough that we don't care about the RAM impact. but if i'm wrong we can change it :woman_shrugging:

obrusvit replied:
I think it's fine. I was just wondering whether there is different reason.

obrusvit committed:
fix(tests): assert multiple in slip39 checklist

mmilata commented on core/src/trezor/ui/__init__.py:
```diff
@@ -100,102 +108,406 @@ def backlight_fade(val: int, delay: int = 14000, step: int = 15) -> None:
-            content_store.append(self.__class__.__name__)
+        self.layout.request_complete_repaint()
+        painted = self.layout.paint()
+        backlight_fade(BacklightLevels.NONE)
```
pushed 3f9975eeec088e1d0ba105293915244fe5776a80, did not do any comprehensive testing but backlight now seems to work normally

obrusvit committed:
fix(tests): fix click tests race condition

mmilata committed:
test(core): fix test_reset_bip39_t2

[no changelog]

mmilata committed:
test(core): fix T3T1 U2F test

[no changelog]

obrusvit commented on core/mocks/generated/trezorui2.pyi:
```diff
@@ -1106,6 +1056,15 @@ class AttachType:
     SWIPE_RIGHT: ClassVar[int]
+
+
+# rust/src/ui/model_tr/layout.rs
```
This will be solved later.

obrusvit committed:
fix(tests): fix T3B1 persistance test

mmilata committed:
test(core): fix T2T1 FIDO2 test

Submodule update only.

[no changelog]

mmilata commented on core/embed/rust/src/ui/layout/base.rs:
```diff
@@ -0,0 +1,84 @@
+}
+
+pub trait Layout<T> {
+    //fn attach(&mut self, ctx: &mut EventCtx, attach_type: AttachType);
```
333c5ec

mmilata commented on core/src/trezor/ui/layouts/homescreen.py:
```diff
@@ -1,8 +1,133 @@
+
+    def __init__(self, layout: Any) -> None:
+        super().__init__(layout=layout)
+        self.context = None
```
333c5ec

matejcik committed:
tests(core): fix upgrade tests

matejcik committed:
style(python): upgrade debuglink.py to 3.10 style type annotations

obrusvit reviewed (approved):
LFG GFL

matejcik committed:
fix(tests): patch over confirm_blob buttonrequest problems

matejcik committed:
feat(core): log ButtonRequests with names for easier debugging

matejcik committed:
feat(core): enforce return type annotations

matejcik committed:
fix(core): patch over problems with confirm_blob_pagination

matejcik committed:
chore: update fixtures

obrusvit committed:
fix(tests): show_multisig title assertions

matejcik committed:
chore: update fixtures for translations

matejcik committed:
chore: update FIDO2 test helpers

obrusvit committed:
fix(tests): remove lang asserts from click slip39

- some non-english texts do not fit on the screen and are paginated,
these tests cannot handle it
- commented out not to block GFL PR, to be resolved later

matejcik committed:
fix(python): do not take legacy snapshots on PinRequest and PassphraseRequest

legacy < 1.11.0 will never answer those because PinRequest and
PassphraseRequest block waiting for PinAck / PassphraseAck over wirelink
and ignore debuglink

romanz commented on core/src/trezor/ui/__init__.py:
```diff
@@ -100,102 +103,429 @@ def backlight_fade(val: int, delay: int = 14000, step: int = 15) -> None:
+            set_current_layout(None)
+            if __debug__:
+                if self.button_request_ack_pending:
+                    raise wire.FirmwareError("button request ack pending")
```
IIUC, it may be happen in case the `ButtonAck` response doesn't arrive before the UI is closed by the user - which may happen due to transport delays (e.g. over BLE).
Is it the expected behaviour?

romanz replied:
CC: @matejcik @M1nd3r

matejcik replied:
> since it only happens in debug builds, probably to make sure `ButtonAck` are correctly sent by the host

this is the answer. in production build we don't want to crash the flow because of it, but there may be some weird consequences.

in an ideal world we'd decouple ButtonRequests from the UI flow by, i dunno, building in a separate ButtonRequest queue in the codec handler, and only sending events to the queue, and the codec itself would resolve whether there are pending BRs, and not e.g. send a next one until the host responds to the previous .... something something. haven't really thought deeply about it.
anyway, the semantic intention is that UI code should *signal* that a ButtonRequest worthy change has happened, but not *require* it.
in fact we planned to start ignoring BRs if the reply doesn't come within 500 ms, which on USB would indicate that the host has crashed

romanz replied:
#6651 will be allow the user to abort/continue in case of a delayed ButtonRequest ACK (depending on the workflow).
